[MIG] l10n_es_ticketbai: Migration to 10.0

xAdrianC-Kernet opened the pull request:
Migración a v10, añadidos los requeriments

acysos committed:
[ADD] l10n_es_aeat_certificate

acysos committed:
[ADD] l10n_es_aeat_soap

[FIX] l10n_es_aeat_certificate: minor fix

oca-travis committed:
[UPD] Update l10n_es_aeat_certificate.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_aeat_certificate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_aeat_certificate/

brendapaniagua committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (29 of 29 strings)

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_aeat_certificate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_aeat_certificate/es/

brendapaniagua committed:
Added translation using Weblate (Catalan)

brendapaniagua committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (29 of 29 strings)

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_aeat_certificate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_aeat_certificate/ca/

ljsalvatierra-binovo committed:
[ADD] l10n_es_ticketbai_api: send invoices to the Basque Country tax agencies (API).
Base addon for l10n_es_ticketbai.
Added Python library dependencies to requirements.txt.

ljsalvatierra-binovo committed:
[ADD] l10n_es_ticketbai_api: send invoices to the Basque Country tax agencies (API).
Base addon for l10n_es_ticketbai.
Added Python library dependencies to requirements.txt.

oca-travis committed:
[UPD] Update l10n_es_ticketbai_api.pot

ljsalvatierra-binovo committed:
[ADD] l10n_es_ticketbai: send invoices to the Basque Country tax agencies.

oca-travis committed:
[UPD] Update l10n_es_ticketbai.pot

ljsalvatierra-binovo committed:
[FIX/I18N] l10n_es_ticketbai: fix invoice with VAT not subject to.
If fiscal position IPSI/IGIC is selected, not subject to cause should be 'RL'.
Update spanish translation.

oca-travis committed:
[UPD] Update l10n_es_ticketbai.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_ticketbai
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_ticketbai/

xAdrianC-Kernet committed:
[MIG] l10n_es_ticketbai_api: Migration to 10.0

xAdrianC-Kernet committed:
[MIG][10.0]complete migration tbai

[MIG][10.0]complete migration tbai

xAdrianC-Kernet commented:
Buenos días @pedrobaeza ,
para finalizar la migración de toda la integración con ticket_bai he migrado en otra rama los módulos l10n_es_ticketbai y l10n_es_aeat_certificate, que necesita el anterior para la gestión de certificados, tengo dudas acerca de cómo proceder con los PR, mi idea era incluir los commits originales más el mío de adecuación a v10 a este mismo PR ya que incluirlo en PR's separados provocará errores al depender de este módulo (l10n_es_ticketbai_api).

¿Sería correcto realizar un PR con la inclusión de los tres módulos o cuál sería el camino más adecuado?

pedrobaeza commented:
Eso depende de los revisores. Normalmente es más cómodo y menos bloqueante un módulo por PR, pero como esto es un backport, tal vez los que quieran revisarlo, prefieren hacerlo todo en bloque. De todas formas, seguir trabajando en 2021 en Python 2.x es totalmente desaconsejable y deberíais enfocar en mi opinión vuestros esfuerzos en migrar a nuevas versiones en lugar de hacer backports...

xAdrianC-Kernet commented:
Totalmente de acuerdo con el uso de python 2.x pero en este caso, cuando se inició el grupo de trabajo nos comprometimos con el resto de empresas de encargarnos de esta parte y queriamos cumplir con ello, además de que por las distintas realidades en cada empresa e implantador hay gente tiene que seguir en estas versiones en algunos casos... Pero como bien dices, la estrategía la estamos enfocando en intentar migrar.

Gracias por la respuesta y me pongo con ello.

blaskurain committed:
[IMP] l10n_es_ticketbai, l10n_es_ticketbai_api Never send messages to IRS from cloned databases, avoid manual commits.

blaskurain committed:
[IMP] l10n_es_ticketbai, l10n_es_ticketbai_api Never send messages to IRS from cloned databases, avoid manual commits.

blaskurain committed:
[REF] l10n_es_ticketbai_api Move XML schema dir to tests and make usage more flexible.

Move the schema directory into the tests directory because after all
that is where it is used and also travis prefers it this way.

Make the usage of the catalog a little more flexible as stepping stone
for related addons.

blaskurain committed:
[IMP] l10n_es_ticketbai_api Update Gipuzkoa Tax Agency Rest Api response codes and URLs

Update according to the latest official documentation ->
https://www.gipuzkoa.eus/documents/2456431/13761128/Anexo+IV.pdf/79d4cafc-eeaf-0b73-d4a9-3eb70c68034a

blaskurain committed:
[IMP] l10n_es_ticketbai_api Update Gipuzkoa Tax Agency Rest Api response codes and URLs

Update according to the latest official documentation ->
https://www.gipuzkoa.eus/documents/2456431/13761128/Anexo+IV.pdf/79d4cafc-eeaf-0b73-d4a9-3eb70c68034a

blaskurain committed:
[IMP] l10n_es_ticketbai Add support for "IVA Exento Repercutido Sujeto"

blaskurain committed:
[FIX] l10n_es_ticketbai_api If an invoice substitutes simplified invoice, it is not a refund

blaskurain committed:
[FIX] l10n_es_ticketbai_api Remove access to tbai.certificate for group_user.

Allowing this access could be a security risk.

blaskurain committed:
[IMP] l10n_es_ticketbai_api Save warning messages from IRS also when invoices get accepted

blaskurain committed:
[IMP] l10n_es_ticketbai Allow the user to say that the line detail is confidential

Gipuzkoa requires that invoices include line details but at the same
time requires that this details should be obfuscated in case the are
to be considered confidential for any reason (medical data for
example).

blaskurain committed:
[IMP] l10n_es_ticketbai Credit notes for invoices created before deploying ticketbai

blaskurain committed:
[IMP] l10n_es_ticketbai, l10n_es_ticketbai_api Add missing tests.

blaskurain committed:
[IMP] l10n_es_ticketbai, l10n_es_ticketbai_api Add missing tests.

blaskurain committed:
[IMP] l10n_es_ticketbai Provide the TBAI testing environment only if available.

The testing environment is not available for all the tax agency. Try
not to lie to the user letting him activate the testing environment
when we know it is not available.

blaskurain committed:
[IMP] l10n_es_ticketbai simplify document schema validation.

Instead of raise + except just ask the schema whether a document is
valid: much simpler and cleaner. Also, keeps travis happy by not
logging assertion errors.

blaskurain committed:
[IMP] l10n_es_ticketbai_api: report error if partner has not country code

oca-travis committed:
[UPD] Update l10n_es_ticketbai.pot

oca-travis committed:
[UPD] Update l10n_es_ticketbai_api.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_ticketbai_api
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_ticketbai_api/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-11.0/l10n-spain-11.0-l10n_es_ticketbai
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-11-0/l10n-spain-11-0-l10n_es_ticketbai/

xAdrianC-Kernet committed:
[FIX/UPD][10.0] migrate commits fundamental functionality from v11 and adapt them to v10

IvanHRN commented on l10n_es_ticketbai/models/account_invoice.py:
```diff
@@ -0,0 +1,511 @@
+            sign = -1
+        else:
+            sign = 1
+        amount_tax = sum(self.invoice_line_tax_ids.mapped('balance'))
```
Debido a que el total de la línea no se calcula correctamente he sustituido esta línea y la siguiente (amount_tax y el return) por el código que se muestra a continuación.

`return "%.2f" % (sign * self._get_price_total())`

La línea llama a un método que lo que hace es devolver los impuestos incluidos. 

```@api.multi
    def _get_price_total(self):
        self.ensure_one()
        currency = self.invoice_id and self.invoice_id.currency_id or None
        price = self.price_unit * (1 - (self.discount or 0.0) / 100.0)
        taxes = False
        if self.invoice_line_tax_ids:
            taxes = self.invoice_line_tax_ids.compute_all(price, currency, self.quantity, product=self.product_id,
                                                          partner=self.invoice_id.partner_id)
        price_subtotal = taxes['total_excluded'] if taxes else self.quantity * price
        return taxes['total_included'] if taxes else price_subtotal```

xAdrianC-Kernet replied:
He estado probando y he incluido estos cambios, la forma es mucho más limpia además de correcta.

IvanHRN commented on l10n_es_ticketbai/views/account_invoice_views.xml:
```diff
@@ -0,0 +1,98 @@
+            <field name="model">account.invoice</field>
+            <field name="inherit_id" ref="account.invoice_form"/>
+            <field name="arch" type="xml">
+                <page name="other_info" position="after">
```
En los casos de en los que se produce una exención del impuesto no se guarda el valor en la línea del impuesto por lo que el servicio de ticketbai devuelve un error.

Para solucionarlo se ha añadido el campo de manera invisible al tree de los impuestos de la factura. El método donde se guardan estos datos es el método del módulo de account "_onchange_invoice_line_ids"

```
               <xpath expr="//field[@name='tax_line_ids']/tree" position="inside">
                    <field name="tbai_vat_exemption_key" invisible="1"/>
                </xpath>

xAdrianC-Kernet replied:
Añadido el campo a la vista

IvanHRN reviewed (commented):
Buenas, 

He realizado alguna pruebas (teniendo en cuenta los cambios comentados) y estoy teniendo problemas con algunas posiciones fiscales como son las de recargo de equivalencia o las de IRPF, las cuales se desglosan en varios impuestos.

Las siguientes posiciones fiscales funcionan correctamente:

- Régimen Nacional
- Regimen Intracomunitario
- Régimen Extracomunitario
- Inversión de sujeto pasivo nacional
- Régimen nacional revendedor con inversión de sujeto pasivo.

ao-landoo committed:
[IMP] Remove l10n_es_account_invoice_sequence module dependency

xAdrianC-Kernet commented:
hola @IvanHRN he estado haciendo las pruebas de todo lo que me comentas, he implementado los cambios del cálculo de los importes de impuestos y el campo del motivo de exención en las líneas de impuestos en el mismo commit de adaptación a la v10.

He estado probando los fallos en:
- Facturas con inversión del sujeto pasivo
- Facturas con ivas exentos

Los errores devueltos por el ticketbai son los mismo que al hacer el mismo proceso con el módulo de la v11, del cual este es un backport, por lo que igual es mejor que validemos lo que hay hasta ahora mergeado en la v11, mergear en la 10 y sobre ello hacer pr para modificar esos warnings, y traerlos o llevarlos a la v11 en un commit aislado.

Al ser warnings y aceptar las facturas igualmente creo que no son tan limitantes, pero comentamos.

IvanHRN reviewed (approved):
Revisados cambios todo correcto. Gracias!!

enriquemartin reviewed (approved):
Salvo los problemas con algunas posiciones fiscales, con las cuales se aceptan las facturas pero con warning, todo correcto.
Estoy de acuerdo en no bloquear este PR por esos warnings; de hecho, se dan en todas las versiones.

xAdrianC-Kernet commented:
hola @pedrobaeza no sé si habría posibilidad de mergear este pr, travis da fallos pero todos ellos relacionados con el módulo l10n_es_account_invoice_sequence, funcionalmente está correcto y tampoco tiene conflictos con master, ya me dices.
Gracias.

pedrobaeza commented:
No, si da fallos es porque el PR estropea algo.
